### ISSUE TITLE
Distribute as a Django app instead of a full project

### ISSUE THREAD

**Issue by johanndt** [inventree/InvenTree#861]
Hi, amazing project!

I'm so glad to see a Django inventory solution. Looking through the source it seems this is a whole stand-alone Django project, and while you have the API, can add plugins and other apps, it's not set up in a way to allow it to be used in other Django projects.

It would be great to list Inventree in requirements.txt, add it to INSTALLED_APPS and then have it work with existing projects. There are a lot of use-cases: An SME might use Django for their e-commerce store, CMS / website, reporting, or some custom application handling other aspects of their business.

Currently a dev can't use Inventree as part of a bigger system without having to maintain a full fork of it with their own changes added on top.

Once it's a module, you get to leverage everything the Django ecosystem has to offer without having to include it yourself. Someone wants 2-factor authentication, they can just add it to their own project. Someone wants a online shopping cart, add one of the dozens already in Django-land. Multi-tenancy, permissions, reporting, integrations to accounting systems - it's all out there.

Is this something you'd entertain and open to receiving PRs for? :-)

Either way, great project.

**Comment by SchrodingersGat** [comment]
Hi,

I'm not familiar with how such a change would be implemented, but I am willing to entertain the idea. Would there be much alteration required to the back-end code? Would the user experience change for any people currently using InvenTree?

**Comment by johanndt** [comment]
It probably doesn't change much at all. I think as a first step all that would be needed is creating the package and including only the files which are needed.

There's a django tutorial: https://docs.djangoproject.com/en/3.0/intro/reusable-apps/

But I've seen other apps still have project files to run as a standalone installation, like InvenTree currently does.

There's sometimes things you can do to make integrations easier, by allowing "hooks" or making it easy to extend functionality, but it's something people can add or contribute to over time as they need it. For example using the "standard" names for certain template blocks:

https://django-reusable-app-docs.readthedocs.io/en/latest/apps/templates.html

And using namespaces for your urls and keeping them in the app folders - these you seem to be doing already. :-)

TL;DR - it shouldn't change much at all to just create a package config and put it out there. People can then use it and contribute as needed.

**Comment by gabsi55** [comment]
A django package would be highly apprciated by me.
I would love to integrate it in our working groups intranet to give the users similar look and feel.

**Comment by SchrodingersGat** [comment]
I'm happy for someone to contribute this as a feature, I can see that it would be useful. I don't have time at the moment to look into implementing it myself, but I'm happy to review.

**Comment by eeintech** [comment]
@johanndt Want to give this a shot? You seemed to have already dug in some information on how to proceed.

**Comment by johanndt** [comment]
Unfortunately my current workload is prohibitive to put it mildy. :sob: 

Sorry.

**Comment by tvanesse** [comment]
I know this is getting old but was there any recent talk on this topic? Are you still open to PRs to make this happen?

We would greatly benefit from having InvenTree released as a reusable app since we already run a big Django project and "only" need to integrate an inventory management system. Having InvenTree running in its own container and relying on the API for integration really feels like a waste of resources. Because both projects (ours and InvenTree) could essentially coexist within the same Django project, avoiding to have 2 servers, 2 workers, 2 dbs, 2 authentication systems, etc. running in parallel. Also avoiding the whole API-based integration logic altogether.

Do you know of anybody who already managed to integrate InvenTree into an existing Django project (i.e., without running InvenTree as a standalone app)?

**Comment by matmair** [comment]
InvenTree is a fast system with a lot of base layer things like logging, user management, tracing, middleware, api permission, plugins, background tasks implemented in opinionated ways.
Extracting that is a lot of work, I do not see the core team spending time on it.
It might seem like it is "just" an inventory system but there is a lot going on to fulfill that role.

If you want to work on this I'd be open to taking a look @tvanesse but this would be a length discussion before we take on that burden

**Comment by tvanesse** [comment]
Hey @matmair , thanks a lot for the prompt response.

I totally understand. After looking at the code for some time, I realise it would indeed take quite a bit of work to make InvenTree truly reusable (djangly speaking). If I were a maintainer, I would also keep it as a monolithic project and advocate for API integration.

I still believe having a reusable inventory management system (as a Django app) is possible, but then it needs to be thought/designed with that goal in mind from day 1.

**Comment by matmair** [comment]
Any inventory system worth its salt will be at least as complex as InvenTree but one can dream
